sips/cluster_consensus.md

```diff
@@ -0,0 +1,330 @@
+	Beacon            BeaconNode
+	OperatorID        OperatorID
+	QBFTParams        QBFTParams
+    
+}
+
+type QBFTParams interface {
+    Operators()        []OperatorID
+    Quorum()           uint64
+    PartialQuorum() 	  uint64
+}
+
+
+type Share struct {
+    // New field:
+	// HighestAttestingSlot holds the highest slot for which attester duty ran for a validator
+    HighestAttestingSlot spec.Slot
+}
+
+// ClusterRunner manages the duty cycle for a certain slot
+type ClusterRunner interface {
+	// Start the duty lifecycle for the given slot. Emits a message.
+    StartDuties(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error
+	// Processes cosensus message
+    ProcessConsensus(consensusMessage *qbft.Message) error
+	// Processes a post-consensus message
+    ProcessPostConsensus(msg PartialSignatureMessages) 
+}
+
+type ClusterRunner struct {
+	Shares          map[ValidatorPubkey]Share
+    QBFTParams     QBFTParams
+	QBFTController *qbft.Controller
+	BeaconNetwork  *types.BeaconNetwork
+}
+
+// BeaconVote is the consensus data
+type BeaconVote struct {
+    BlockRoot         phase0.Root
+    Source            phase0.Checkpoint
+    Target            phase0.Checkpoint
+}
+
+func ConstructAttestation(vote BeaconVote, duty AttesterDuty) Attestation {
+    bits := bitfield.New(duty.CommitteeLength)
+    bits.Set(duty.ValidatorCommitteeIndex)
+
+    return Attestation{
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        Source: vote.Source,
+        Target: vote.Target,
+        CommiteeIndex: duty.CommitteeIndex,
+        AggregationBits: bits,
+    }
+}
+
+func ConstructSyncCommittee(vote BeaconVote, duty AttesterDuty) SyncCommitteeMessage{
+    return SyncCommitteeMessage
+    {
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        ValidatorIndex: duty.ValidatorIndex,
+    }
+}
+```
+#### PartialSignatureMessages
+
+The current structure that we have in code can be unchanged.
+
+```go
+type PartialSignatureMessages struct {
+	Type     PartialSigMsgType
+	Slot     phase0.Slot
+	Messages []*PartialSignatureMessage
+}
+
+// PartialSignatureMessage is a msg for partial Beacon chain related signatures
+type PartialSignatureMessage struct {
+	PartialSignature Signature `ssz-size:"96"` // The Beacon chain partial Signature for a duty
+	SigningRoot      [32]byte  `ssz-size:"32"` // the root signed in PartialSignature
+    Signer           OperatorID // TODO: there is an optimization where we don't have to duplicate this field
+}
+```
+
+We must have the following flow:
+1. When you `calculateExpectedRootsAndDomain` (similar to `sync_committee_aggregator`). Use the `duty` objects to reconstruct the proper beacon data objects (i.e. `Attestation`).
+2. The above calculation can be used to create a mapping of `ValidatorIndex` to `root`
+3. When processing the messages find all the roots that have quorums, mark them, and sumbit corresponding beacon data to beacon chain.
```

Comment:
So, for attestation duties, we can relate the validators to the signing root. But what about the signatures for the sync committee duties? The signing root should be the same if I'm not mistaken.

Reply:
It isn't the same because we sign Beacon Objects

Reply:
https://github.com/bloxapp/ssv-spec/blob/ebb60143f17ea25edf09edeb398559d764b0e01b/ssv/sync_committee.go#L138-L143

Reply:
you sign the head root for sync committee
for attestation it is the head root + source + target